fs: nvs: Allow the application to better control the garbage collector triggering

RICCIARDI-Adrien opened the pull request:
The goal is to make sure a determined amount of contiguous free space is present into the NVS for a subsequent write.

The subsequent write needs to write to the NVS as fast as possible, so triggering the NVS garbage collector must be avoided.

This is the reason why this function is introduced. If the GC must be triggered to make some room, it will trigger it, so the subsequent write will not be slowed by the GC process.

If this feature is OK for the maintainers, I can of course add Twister tests and documentation if required.

Laczen commented:
I am not a maintainer, so I can't decide if this is OK.

Feedback on this proposal: It would be good to know if data of certain size can be written without triggering GC, however given that zephyr is a multithread solution this is only valid immediately after exiting such a routine. When the write is triggered this infomation can already be changed because another thread has written data, still resulting in a need to perform gc.

RICCIARDI-Adrien commented:
Your feedback is more than welcome !
Indeed, this function is intended to be used in a specific configuration, with a single thread accessing the NVS during the whole life of the application.
Do you have suggestions on how to better name this ? Would the implementation be acceptable ?

Laczen commented:
> Your feedback is more than welcome ! Indeed, this function is intended to be used in a specific configuration, with a single thread accessing the NVS during the whole life of the application. Do you have suggestions on how to better name this ? Would the implementation be acceptable ?

The use case seems to be very limited to add this method. NVS has been designed to try and avoid any unnecessary erases, the proposed change violates this principle. It is not up to me to decide if this can be added, so I'm removing my assignment.

RICCIARDI-Adrien commented:
Hi @Laczen, maybe a more generic approach would be better. The API would be based on 3 functions :
* A first function like `nvs_get_curr_sect_free_space()` would return something like `fs->ate_wra - fs->data_wra - nvs_al_size(fs, sizeof(struct nvs_ate)`, so the application can know how many contiguous free space remains in the current sector.
* A second function like `nvs_compact_next_sect()` would close the current sector and invoke the GC on the next sector.
* A third function like `nvs_start_compaction()` would initialize an internal counter of the garbage collected sectors, to avoid calling `nvs_compact_next_sect()` endlessly.

This way, an application needing to know if there is enough room to write a chunk without incurring the GC delay could run a simple loop like :
```
uint32_t data_sz = <size of chunk to write>;
ssize_t free_sz, ret;
nvs_start_compaction(&fs);
do {
	free_sz = nvs_compact_next_sect(&fs);
	if (free_sz < 0) {
		/* Handle error */
	}

	if (free_sz < data_sz) {
		ret = nvs_compact_next_sect(&fs);
		if (ret) {
			/* Handle error */
			/* nvs_compact_next_sect() will return something like -ELOOP when all sectors have been GC */
		}
	} 
} while (free_sz < data_sz);
```

At the end of this loop, the application knows whether `free_sz` is lower, equal or greater to `data_sz`.

This case is needed for several battery-powered devices that do not have the time to run a whole GC procedure during some critical power-off operation, so it can "make room" in advance in the NVS before reaching the critical operation. In such case, the application would make sure to not have multiple threads writing to the NVS at the same time, otherwise all this would be useless.

RICCIARDI-Adrien commented:
OK.
Would the `nvs_sector_start_next()` also trigger compaction ? For instance, if `nvs_sector_max_dsize()` is called on the last opened sector and there is not enough room for the application, there should be a way other than calling `nvs_write()` to trigger the GC, then `nvs_sector_max_dsize()` could be called again to see if more room is available.

Laczen commented:
> OK. Would the `nvs_sector_start_next()` also trigger compaction ? For instance, if `nvs_sector_max_dsize()` is called on the last opened sector and there is not enough room for the application, there should be a way other than calling `nvs_write()` to trigger the GC, then `nvs_sector_max_dsize()` could be called again to see if more room is available.

Yes, maybe 'nvs_sector_start_next' should be replaced by 'nvs_sector_use_next' to avoid confusion. The routine closes the present sector and calls GC.

RICCIARDI-Adrien committed:
fs: nvs: Allow application to switch sector to get free space

Add an API function allowing the application to determine the amount of
free bytes in the current sector.

Add a second API function allowing the application to switch to the next
NVS sector, calling the garbage collector on such sector.

The goal is togive more granularity to the application to control when
the garbage collector is triggered.

Signed-off-by: Adrien Ricciardi <aricciardi@baylibre.com>

RICCIARDI-Adrien commented:
@Laczen What do you think of this new implementation ?

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -1346,3 +1346,31 @@ ssize_t nvs_calc_free_space(struct nvs_fs *fs)
+
+	ret = nvs_sector_close(fs);
+	if (ret) {
+		return ret;
```
`if (ret != 0)` or `if (ret < 0)`

RICCIARDI-Adrien replied:
Changed in `if (ret != 0)`.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -1346,3 +1346,31 @@ ssize_t nvs_calc_free_space(struct nvs_fs *fs)
+int nvs_sector_use_next(struct nvs_fs *fs)
+{
+	int ret;
+
```
needs locking of fs.

RICCIARDI-Adrien replied:
Nice spot, fixed !

Laczen commented on include/zephyr/fs/nvs.h:
```diff
@@ -164,6 +164,26 @@ ssize_t nvs_read_hist(struct nvs_fs *fs, uint16_t id, void *data, size_t len, ui
+/**
+ * @brief Close the currently active sector and switch to the next one.
+ *
+ * @note The garbage collector is called on the new sector.
```
This should have a warning, something like "this routine breaks the aim of nvs to avoid any unnecessary flash erases, using this routine extensively can result in premature failure of the flash device".

RICCIARDI-Adrien replied:
Warning added.

Laczen reviewed (dismissed):
LGTM, maybe a test should be added.

de-nordic commented:
@rghaddab Can you also take a look?

rghaddab commented on subsys/fs/nvs/nvs.c:
```diff
@@ -1346,3 +1346,40 @@ ssize_t nvs_calc_free_space(struct nvs_fs *fs)
+
+	ate_size = nvs_al_size(fs, sizeof(struct nvs_ate));
+
+	return fs->ate_wra - fs->data_wra - ate_size;
```
we should take into account as well the NVS_DATA_CRC_SIZE, right ?

RICCIARDI-Adrien replied:
Indeed, this is a nice spot !
Fixed.
